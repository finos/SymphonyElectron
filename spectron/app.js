const Application = require('spectron').Application;
const path = require('path');
const chai = require('chai');
const chaiAsPromised = require('chai-as-promised');
const chaiRoughly = require('chai-roughly');

global.before(function () {
    chai.should();
    chai.use(chaiAsPromised);
    chai.use(chaiRoughly);
});

class App {

    constructor(options) {

        this.options = options;

        if (!this.options.path){
            this.options.path = App.getAppPath();
        }

        this.app = new Application(this.options);
    }

    startApplication() {
        return this.app.start().then(() => {
            chaiAsPromised.transferPromiseness = this.app.transferPromiseness;
            return this.app
        });
    }

    static getAppPath() {
        let electronPath = path.join(__dirname, '..', 'dist/mac/Symphony.app/Contents/MacOS/Symphony');
        if (process.platform === 'win32') {
            electronPath += '.cmd';
        }
        return electronPath
    }

}

module.exports = App;