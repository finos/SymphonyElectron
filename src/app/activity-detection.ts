import { app, powerMonitor, WebContents } from 'electron';
import Timer = NodeJS.Timer;

import { logger } from '../common/logger';
import { windowHandler } from './window-handler';

class ActivityDetection {
  private idleThreshold: number;
  private window: WebContents | null;
  private timer: Timer | undefined;
  private queryInterval: NodeJS.Timer | undefined;

  constructor() {
    this.idleThreshold = 60 * 60 * 1000;
    this.window = null;
  }

  /**
   * Sets the window and the idle threshold from the web app
   *
   * @param window {Electron.BrowserWindow}
   * @param idleThreshold {number}
   */
  public setWindowAndThreshold(
    window: WebContents,
    idleThreshold: number,
  ): void {
    this.window = window;
    this.idleThreshold = idleThreshold;
    if (this.queryInterval) {
      clearInterval(this.queryInterval);
    }
    this.startActivityMonitor();
    logger.info(
      `activity-detection: Initialized activity detection with an idleThreshold of ${idleThreshold}`,
    );
  }

  /**
   * Start a timer for monitoring user active
   */
  private startActivityMonitor(): void {
    if (app.isReady()) {
      logger.info(`activity-detection: Starting activity monitor`);
      this.queryInterval = setInterval(() => {
        const idleTime = powerMonitor.getSystemIdleTime();
        this.activity(idleTime);
      }, this.idleThreshold);
    }
  }

  /**
   * Validates and send the user activity based on
   * the idle threshold set be the web app
   *
   * @param idleTime {number}
   */
  private activity(idleTime: number): void {
    const idleTimeInMillis = idleTime * 1000;
    if (idleTimeInMillis < this.idleThreshold) {
      this.sendActivity(idleTimeInMillis);
      if (this.timer) {
        clearInterval(this.timer);
      }
      // set auto reload to false so the
      // activate func works normally
      windowHandler.setIsAutoReload(false);
      this.timer = undefined;
      logger.info(
        `activity-detection: activity occurred, updating the client!`,
      );
      return;
    }

    if (!this.timer) {
      logger.info(
        `activity-detection: user is inactive, started monitoring for every 1 sec`,
      );
      // starts monitoring for user activity every 1 sec
      // when user goes inactive
      this.timer = setInterval(() => {
        if (app.isReady()) {
          const activeTime = powerMonitor.getSystemIdleTime();
          this.activity(activeTime);
        }
      }, 1000);
    }
  }

  /**
   * Send user activity to the web app
   *
   * @param idleTime {number}
   */
  private sendActivity(idleTime: number): void {
    if (this.window && !this.window.isDestroyed()) {
      this.window.send('activity', idleTime || 1);
      logger.info(`activity-detection: Sending activity status to the client!`);
    }
  }
}

const activityDetection = new ActivityDetection();

export { activityDetection };
