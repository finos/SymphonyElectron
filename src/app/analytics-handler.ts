import { WebContents } from 'electron';

export interface IAnalyticsData {
  element: AnalyticsElements;
  action_type?:
    | MenuActionTypes
    | ScreenSnippetActionTypes
    | ToastNotificationActionTypes;
  action_result?: AnalyticsActions;
  extra_data?: object;
}

export interface ICrashData extends IAnalyticsData {
  process: SDACrashProcess;
  crashCause: string;
  windowName: string;
  miniDump?: string;
}

export enum MenuActionTypes {
  AUTO_LAUNCH_ON_START_UP = 'auto_launch_on_start_up',
  ALWAYS_ON_TOP = 'always_on_top',
  MINIMIZE_ON_CLOSE = 'minimize_on_close',
  FLASH_NOTIFICATION_IN_TASK_BAR = 'flash_notification_in_task_bar',
  HAMBURGER_MENU = 'hamburger_menu',
  REFRESH_APP_IN_IDLE = 'refresh_app_in_idle',
}

export enum ScreenSnippetActionTypes {
  SCREENSHOT_TAKEN = 'screenshot_taken',
  ANNOTATE_ADDED_PEN = 'annotate_added_pen',
  ANNOTATE_ADDED_HIGHLIGHT = 'annotate_added_highlight',
  ANNOTATE_DONE = 'annotate_done',
  ANNOTATE_CLEARED = 'annotate_cleared',
  ANNOTATE_ERASED = 'annotate_erased',
}

export enum ToastNotificationActionTypes {
  TOAST_CLOSED = 'toast_closed',
}

export enum AnalyticsActions {
  ENABLED = 'ON',
  DISABLED = 'OFF',
}

export enum AnalyticsElements {
  MENU = 'Menu',
  SCREEN_CAPTURE_ANNOTATE = 'screen_capture_annotate',
  TOAST_NOTIFICATION = 'toast_notification',
  SDA_CRASH = 'sda_crash',
}

export enum SDACrashProcess {
  MAIN = 'main',
  RENDERER = 'renderer',
  GPU = 'gpu',
}

const MAX_EVENT_QUEUE_LENGTH = 50;
const analyticsCallback = 'analytics-callback';

class Analytics {
  private preloadWindow: WebContents | undefined;
  private analyticsEventQueue: IAnalyticsData[] = [];

  /**
   * Stores the reference to the preload window
   *
   * @param webContents {WeContents}
   */
  public registerPreloadWindow(webContents: WebContents): void {
    this.preloadWindow = webContents;

    if (!(this.preloadWindow && !this.preloadWindow.isDestroyed())) {
      return;
    }
    if (this.analyticsEventQueue && this.analyticsEventQueue.length > 0) {
      this.analyticsEventQueue.forEach((eventData) => {
        if (this.preloadWindow && !this.preloadWindow.isDestroyed()) {
          if (eventData.element === AnalyticsElements.SDA_CRASH) {
            eventData = eventData as ICrashData;
          }
          this.preloadWindow.send(analyticsCallback, eventData);
        }
      });
      this.resetAnalytics();
    }
  }

  /**
   * Sends the analytics events to the web client
   *
   * @param eventData {IAnalyticsData}
   */
  public track(eventData: IAnalyticsData): void {
    if (eventData.element === AnalyticsElements.SDA_CRASH) {
      eventData = eventData as ICrashData;
    }
    if (this.preloadWindow && !this.preloadWindow.isDestroyed()) {
      this.preloadWindow.send(analyticsCallback, eventData);
      return;
    }
    this.analyticsEventQueue.push(eventData);
    // don't store more than specified limit. keep most recent events.
    if (this.analyticsEventQueue.length > MAX_EVENT_QUEUE_LENGTH) {
      this.analyticsEventQueue.shift();
    }
  }

  /**
   * Clears the analytics queue
   */
  public resetAnalytics(): void {
    this.analyticsEventQueue = [];
  }
}

const analytics = new Analytics();

export { analytics };
