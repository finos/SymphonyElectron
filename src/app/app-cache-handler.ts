import { app, session } from 'electron';
import * as fs from 'fs';
import * as path from 'path';

import { logger } from '../common/logger';

// Cache check file path
const cacheCheckFilePath: string = path.join(app.getPath('userData'), 'CacheCheck');

/**
 * Deletes app cache file if exists or clears
 * the cache for the session
 */
export const cleanUpAppCache = async (): Promise<void> => {
    if (fs.existsSync(cacheCheckFilePath)) {
        await fs.unlinkSync(cacheCheckFilePath);
        logger.info(`app-cache-handler: last exit was clean, deleted the app cache file`);
        return;
    }
    await new Promise((resolve) => session.defaultSession ? session.defaultSession.clearCache(resolve) : null);
    logger.info(`app-cache-handler: we didn't have a clean exit last time, so, cleared the cache that may have been corrupted!`);
};

/**
 * Creates a new file cache file on app exit
 */
export const createAppCacheFile = (): void => {
    logger.info(`app-cache-handler: this is a clean exit, creating app cache file`);
    fs.writeFileSync(cacheCheckFilePath, '');
};
