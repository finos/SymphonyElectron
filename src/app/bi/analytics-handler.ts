import { app, WebContents } from 'electron';
import * as fs from 'fs';
import * as path from 'path';
import { logger } from '../../common/logger';

export interface IAnalyticsData {
  element: AnalyticsElements;
  action_type?:
    | MenuActionTypes
    | ScreenSnippetActionTypes
    | ToastNotificationActionTypes
    | SDAUserSessionActionTypes
    | InstallActionTypes;
  action_result?: AnalyticsActions;
  extra_data?: object;
}

export interface ICrashData extends IAnalyticsData {
  process: SDACrashProcess;
  crashCause: string;
  windowName: string;
  miniDump?: string;
}

export interface ISessionData extends IAnalyticsData {
  extra_data?: {
    sessionStartDatetime?: string;
    machineStartDatetime?: string;
    machineId?: string;
    InstallVariant?: string;
    osName?: string;
    osVersion?: string;
    osLanguage?: string;
    osTimeZone?: string;
    cpuNumberOfCores?: number;
    cpuMaxFrequency?: number;
    cpuUsagePercent?: number;
    maxCPUUsagePercent?: number;
    memoryTotal?: number;
    memoryUsedPercent?: number;
    maxMemoryUsedPercent?: number;
    sdaUsedMemory?: number;
    memoryAvailable?: number;
    vdi?: boolean;
    endReason?: string;
    crashProcess?: string;
  };
}

export interface IInstallData extends IAnalyticsData {
  extra_data?: {
    installLocation: string;
    installType: string;
  };
}

export enum MenuActionTypes {
  AUTO_LAUNCH_ON_START_UP = 'auto_launch_on_start_up',
  ALWAYS_ON_TOP = 'always_on_top',
  MINIMIZE_ON_CLOSE = 'minimize_on_close',
  FLASH_NOTIFICATION_IN_TASK_BAR = 'flash_notification_in_task_bar',
  HAMBURGER_MENU = 'hamburger_menu',
  REFRESH_APP_IN_IDLE = 'refresh_app_in_idle',
}

export enum ScreenSnippetActionTypes {
  SCREENSHOT_TAKEN = 'screenshot_taken',
  ANNOTATE_ADDED_PEN = 'annotate_added_pen',
  ANNOTATE_ADDED_HIGHLIGHT = 'annotate_added_highlight',
  ANNOTATE_ADD = 'annotate_done',
  ANNOTATE_CLEARED = 'annotate_cleared',
  ANNOTATE_ERASED = 'annotate_erased',
  ANNOTATE_COPY = 'annotate_copy',
  ANNOTATE_SAVE_AS = 'annotate_save_as',
  ANNOTATE_CLOSE = 'annotate_close',
}

export enum ToastNotificationActionTypes {
  TOAST_CLOSED = 'toast_closed',
}

export enum SDAUserSessionActionTypes {
  Start = 'Start',
  End = 'End',
  Login = 'Login',
  Logout = 'Logout',
  Crash = 'Crash',
  ForceReload = 'Force_reload',
}

export enum InstallActionTypes {
  InstallStarted = 'Install_started',
  InstallCompleted = 'Install_completed',
  InstallFailed = 'Install_failed',
}

export enum InstallTypes {
  Auto = 'auto',
  Manual = 'manual',
}

export enum InstallLocationTypes {
  PROG_FILES = 'PROG_FILES',
  REMOTE = 'REMOTE',
  LOCAL = 'LOCAL',
  CUSTOM = 'CUSTOM',
}

export enum SDAEndReasonTypes {
  Reboot = 'Reboot',
  Closed = 'Closed',
  Crashed = 'Crashed',
}

export enum AnalyticsActions {
  ENABLED = 'ON',
  DISABLED = 'OFF',
}

export enum AnalyticsElements {
  MENU = 'Menu',
  SCREEN_CAPTURE_ANNOTATE = 'screen_capture_annotate',
  TOAST_NOTIFICATION = 'toast_notification',
  SDA_CRASH = 'sda_crash',
  SDA_SESSION = 'sda_session',
  SDA_INSTALL = 'sda_install',
}

export enum SDACrashProcess {
  MAIN = 'main',
  RENDERER = 'renderer',
  GPU = 'gpu',
}

const MAX_EVENT_QUEUE_LENGTH = 50;
const analyticsCallback = 'analytics-callback';

class Analytics {
  private preloadWindow: WebContents | undefined;
  private analyticsEventQueue: IAnalyticsData[] = [];
  private analyticsEndEventQueue: IAnalyticsData[] = [];
  private analyticEventsDataFilePath = path.join(
    app.getPath('userData'),
    'analytics.json',
  );

  /**
   * Stores the reference to the preload window
   *
   * @param webContents {WeContents}
   */
  public registerPreloadWindow(webContents: WebContents): void {
    this.preloadWindow = webContents;

    if (!(this.preloadWindow && !this.preloadWindow.isDestroyed())) {
      return;
    }
    if (this.analyticsEventQueue && this.analyticsEventQueue.length > 0) {
      this.analyticsEventQueue.forEach((eventData) => {
        if (this.preloadWindow && !this.preloadWindow.isDestroyed()) {
          if (eventData.element === AnalyticsElements.SDA_CRASH) {
            eventData = eventData as ICrashData;
          }
          this.preloadWindow.send(analyticsCallback, eventData);
        }
      });
      this.resetAnalytics();
    }
  }

  /**
   * Sends the analytics events to the web client
   *
   * @param eventData {IAnalyticsData}
   */
  public track(eventData: IAnalyticsData): void {
    if (eventData.element === AnalyticsElements.SDA_CRASH) {
      eventData = eventData as ICrashData;
    }
    // Store these events and send them in the next SDA session
    if (
      eventData.action_type === SDAUserSessionActionTypes.End ||
      eventData.action_type === SDAUserSessionActionTypes.Logout ||
      eventData.action_type === InstallActionTypes.InstallStarted
    ) {
      this.analyticsEndEventQueue.push(eventData);
      return;
    }
    this.sendAnalyticsOrAddToQueue(eventData);
  }

  /**
   * Writes all the pending stats into a file
   */
  public writeAnalyticFile = () => {
    try {
      fs.writeFileSync(
        this.analyticEventsDataFilePath,
        JSON.stringify(this.analyticsEndEventQueue, null, 2),
        { encoding: 'utf8' },
      );
      logger.info(
        `analytics-handler: updated analytic values with the data ${JSON.stringify(
          this.analyticsEndEventQueue,
        )}`,
      );
    } catch (error) {
      logger.error(
        `analytics-handler: failed to update analytic with ${JSON.stringify(
          this.analyticsEndEventQueue,
        )}`,
        error,
      );
    }
  };

  /**
   * Sends all the locally stored stats
   */
  public sendLocalAnalytics = () => {
    if (fs.existsSync(this.analyticEventsDataFilePath)) {
      const localStats = fs.readFileSync(
        this.analyticEventsDataFilePath,
        'utf8',
      );
      if (!localStats) {
        return;
      }
      let parsedStats: ISessionData[];
      try {
        parsedStats = JSON.parse(localStats);
        logger.info(
          `analytics-handler: parsed stats JSON file with data`,
          parsedStats,
        );
        if (parsedStats && parsedStats.length) {
          parsedStats.forEach((event) => {
            this.sendAnalyticsOrAddToQueue(event);
          });
          fs.unlinkSync(this.analyticEventsDataFilePath);
        }
      } catch (e: any) {
        logger.error(
          `analytics-handler: parsing stats JSON file failed due to error ${e}`,
        );
      }
    }
    if (this.analyticsEndEventQueue.length > 0) {
      this.analyticsEndEventQueue.forEach((eventData) => {
        this.sendAnalyticsOrAddToQueue(eventData);
      });
      this.analyticsEndEventQueue = [];
    }
  };

  /**
   * Clears the analytics queue
   */
  public resetAnalytics(): void {
    this.analyticsEventQueue = [];
  }

  private sendAnalyticsOrAddToQueue = (eventData) => {
    if (this.preloadWindow && !this.preloadWindow.isDestroyed()) {
      this.preloadWindow.send(analyticsCallback, eventData);
      return;
    }
    this.analyticsEventQueue.push(eventData);
    // don't store more than specified limit. keep most recent events.
    if (this.analyticsEventQueue.length > MAX_EVENT_QUEUE_LENGTH) {
      this.analyticsEventQueue.shift();
    }
  };
}

const analytics = new Analytics();

export { analytics };
