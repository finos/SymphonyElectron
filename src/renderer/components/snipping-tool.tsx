import { ipcRenderer } from 'electron';
import * as React from 'react';
import { i18n } from '../../common/i18n-preload';
import { analytics, AnalyticsElements, ScreenSnippetActionTypes } from './../../app/analytics-handler';
import AnnotateArea from './annotate-area';
import ColorPickerPill, { IColor } from './color-picker-pill';

const { useState, useRef, useEffect, useLayoutEffect } = React;

export enum Tool {
  pen = 'PEN',
  highlight = 'HIGHLIGHT',
  eraser = 'ERASER',
}

export interface IPath {
  points: IPoint[];
  color: string;
  strokeWidth: number;
  shouldShow: boolean;
  key: string;
}

export interface IPoint {
  x: number;
  y: number;
}

export interface IDimensions {
  width: number;
  height: number;
}

export interface ISnippingToolProps {
  existingPaths?: IPath[];
}

const availablePenColors: IColor[] = [
  { rgbaColor: 'rgba(0, 0, 40, 1)' },
  { rgbaColor: 'rgba(0, 142, 255, 1)' },
  { rgbaColor: 'rgba(38, 196, 58, 1)' },
  { rgbaColor: 'rgba(246, 178, 2, 1)' },
  { rgbaColor: 'rgba(255, 255, 255, 1)', outline: 'rgba(207, 208, 210, 1)' },
];
const availableHighlightColors: IColor[] = [
  { rgbaColor: 'rgba(0, 142, 255, 0.64)' },
  { rgbaColor: 'rgba(38, 196, 58, 0.64)' },
  { rgbaColor: 'rgba(246, 178, 2, 0.64)' },
  { rgbaColor: 'rgba(233, 0, 0, 0.64)' },
];
const SNIPPING_TOOL_NAMESPACE = 'ScreenSnippet';

const SnippingTool: React.FunctionComponent<ISnippingToolProps> = ({ existingPaths }) => {
  // State preparation functions

  const [screenSnippetPath, setScreenSnippetPath] = useState('');
  const [imageDimensions, setImageDimensions] = useState({
    height: 600,
    width: 800,
  });
  const [annotateAreaDimensions, setAnnotateAreaDimensions] = useState({
    height: 600,
    width: 800,
  });
  const [paths, setPaths] = useState<IPath[]>(existingPaths || []);
  const [chosenTool, setChosenTool] = useState(Tool.pen);
  const [penColor, setPenColor] = useState<IColor>({ rgbaColor: 'rgba(0, 142, 255, 1)' });
  const [highlightColor, setHighlightColor] = useState<IColor>(
    { rgbaColor: 'rgba(0, 142, 255, 0.64)' },
  );
  const [
    shouldRenderHighlightColorPicker,
    setShouldRenderHighlightColorPicker,
  ] = useState(false);
  const [shouldRenderPenColorPicker, setShouldRenderPenColorPicker] = useState(
    false,
  );

  const getSnipImageData = ({ }, {
    snipImage,
    annotateAreaHeight,
    annotateAreaWidth,
    snippetImageHeight,
    snippetImageWidth,
  }) => {
    setScreenSnippetPath(snipImage);
    setImageDimensions({ height: snippetImageHeight, width: snippetImageWidth });
    setAnnotateAreaDimensions({ height: annotateAreaHeight, width: annotateAreaWidth });
  };

  useLayoutEffect(() => {
  ipcRenderer.once('snipping-tool-data', getSnipImageData);
    analytics.track({
      element: AnalyticsElements.SCREEN_SNIPPET,
      action_type: ScreenSnippetActionTypes.CAPTURE_TAKEN,
    });
  return () => {
    ipcRenderer.removeListener('snipping-tool-data', getSnipImageData);
  };
  }, []);

  // Hook that alerts clicks outside of the passed refs
  const useClickOutsideExaminer = (
    colorPickerRf: React.RefObject<HTMLDivElement>,
    penRf: React.RefObject<HTMLButtonElement>,
    highlightRf: React.RefObject<HTMLButtonElement>,
  ) => {
    useEffect(() => {
      const handleClickOutside = (event: MouseEvent) => {
        if (
          !colorPickerRf?.current?.contains(event.target as Node) &&
          !penRf?.current?.contains(event.target as Node) &&
          !highlightRf?.current?.contains(event.target as Node)
        ) {
          setShouldRenderHighlightColorPicker(false);
          setShouldRenderPenColorPicker(false);
        }
      };

      document.addEventListener('mousedown', handleClickOutside);
      return () => {
        document.removeEventListener('mousedown', handleClickOutside);
      };
    }, [colorPickerRf, penRf, highlightRf]);
  };

  const colorPickerRef = useRef<HTMLDivElement>(null);
  const penRef = useRef<HTMLButtonElement>(null);
  const highlightRef = useRef<HTMLButtonElement>(null);
  useClickOutsideExaminer(colorPickerRef, penRef, highlightRef);

  // State mutating functions

  const penColorChosen = (color: IColor) => {
    setPenColor(color);
    setShouldRenderPenColorPicker(false);
  };

  const highlightColorChosen = (color: IColor) => {
    setHighlightColor(color);
    setShouldRenderHighlightColorPicker(false);
  };

  const usePen = () => {
    setChosenTool(Tool.pen);
    setShouldRenderPenColorPicker(!shouldRenderPenColorPicker);
    setShouldRenderHighlightColorPicker(false);
  };

  const useHighlight = () => {
    setChosenTool(Tool.highlight);
    setShouldRenderHighlightColorPicker(!shouldRenderHighlightColorPicker);
    setShouldRenderPenColorPicker(false);
  };

  const useEraser = () => {
    setChosenTool(Tool.eraser);
  };

  const clear = () => {
    const updPaths = [...paths];
    updPaths.map((p) => {
      p.shouldShow = false;
      return p;
    });
    setPaths(updPaths);
  };

  // Utility functions

  const getBase64PngData = () => {
    const canvas = document.createElement('canvas');
    canvas.width = imageDimensions.width;
    canvas.height = imageDimensions.height;

    // Creates an in memory canvas for mounting img data without adding it to the DOM
    const ctx = canvas?.getContext('2d') as CanvasRenderingContext2D;

    if (!ctx) {
      // Will only be the case in headless browsers, such as with unit tests
      return 'NO CANVAS';
    }

    const backgroundImage = document.getElementById('backgroundImage') as HTMLImageElement;

    // Fast lane in case there is no drawn SVG paths
    if (paths.length === 0) {
      ctx.drawImage(backgroundImage, 0, 0);
      // Extracts base 64 png img data from the canvas
      const data = canvas.toDataURL('image/png');
      return data;
    }

    // Creates an in memory img without adding it to the DOM
    const img = document.createElement('img');

    const svg = document.getElementById('annotate-area') as HTMLImageElement;
    // Parses SVG image to XML data
    const svgData = new XMLSerializer().serializeToString(svg);
    // Adds the extracted XML data to the in memory img
    img.setAttribute(
      'src',
      'data:image/svg+xml;base64,' + btoa(unescape(encodeURIComponent(svgData))),
    );

    return new Promise((resolve, reject) => {
      // Listens to when the img is loaded in memory and adds the data from the SVG paths + screenSnippet to the canvas
      img.onload = () => {
        ctx.drawImage(backgroundImage, 0, 0);
        ctx.drawImage(img, 0, 0);
        try {
          // Extracts base 64 png img data from the canvas
          const data = canvas.toDataURL('image/png');
          resolve(data);
        } catch (e) {
          reject(e);
        }
      };
    });
  };

  const markChosenColor = (colors: IColor[], chosenColor: string) => {
    return colors.map((color) => {
      if (color.rgbaColor === chosenColor) {
        return { ...color, chosen: true };
      } else {
        return color;
      }
    });
  };

  const getBorderStyle = (tool: Tool) => {
    if (chosenTool !== tool) {
      return undefined;
    }
    if (chosenTool === Tool.pen) {
      const color = penColor.outline ? penColor.outline : penColor.rgbaColor;
      return { border: '2px solid ' + color };
    } else if (chosenTool === Tool.highlight) {
      const color = highlightColor.outline ? highlightColor.outline : highlightColor.rgbaColor;
      return { border: '2px solid ' + color };
    } else if (chosenTool === Tool.eraser) {
      return { border: '2px solid #008EFF' };
    }
    return undefined;
  };

  const done = async () => {
    const base64PngData = await getBase64PngData();
    analytics.track({
      element: AnalyticsElements.SCREEN_SNIPPET,
      action_type: ScreenSnippetActionTypes.CAPTURE_SENT,
    });
    ipcRenderer.send('upload-snippet', { screenSnippetPath, base64PngData });
  };

  return (
    <div className='SnippingTool' lang={i18n.getLocale()}>
      <header>
        <div className='DrawActions'>
          <button
            data-testid='pen-button'
            style={getBorderStyle(Tool.pen)}
            className='ActionButton'
            onClick={usePen}
            title={i18n.t('Pen', SNIPPING_TOOL_NAMESPACE)()}
          >
            <img src='../renderer/assets/snip-draw.svg' />
          </button>
          <button
            data-testid='highlight-button'
            style={getBorderStyle(Tool.highlight)}
            className='ActionButton'
            onClick={useHighlight}
            title={i18n.t('Highlight', SNIPPING_TOOL_NAMESPACE)()}
          >
            <img src='../renderer/assets/snip-highlight.svg' />
          </button>
          <button
            data-testid='erase-button'
            style={getBorderStyle(Tool.eraser)}
            className='ActionButton'
            onClick={useEraser}
            title={i18n.t('Erase', SNIPPING_TOOL_NAMESPACE)()}
          >
            <img src='../renderer/assets/snip-erase.svg' />
          </button>
        </div>
        <div className='ClearActions'>
          <button
            data-testid='clear-button'
            className='ClearButton'
            onClick={clear}
          >
            {i18n.t('Clear', SNIPPING_TOOL_NAMESPACE)()}
          </button>
        </div>
      </header>

      {
        shouldRenderPenColorPicker && (
          <div style={{ marginTop: '64px', position: 'absolute', left: '50%' }} ref={colorPickerRef}>
            <div style={{ position: 'relative', left: '-50%' }}>
              <ColorPickerPill
                data-testid='pen-colorpicker'
                availableColors={markChosenColor(availablePenColors, penColor.rgbaColor)}
                onChange={penColorChosen}
              />
            </div>
          </div>
        )
      }
      {
        shouldRenderHighlightColorPicker && (
          <div style={{ marginTop: '64px', position: 'absolute', left: '50%' }} ref={colorPickerRef}>
            <div style={{ position: 'relative', left: '-50%' }}>
              <ColorPickerPill
                data-testid='highlight-colorpicker'
                availableColors={markChosenColor(
                  availableHighlightColors,
                  highlightColor.rgbaColor,
                )}
                onChange={highlightColorChosen}
              />
            </div>
          </div>
        )
      }

      <main>
        <div className='imageContainer'>
          <AnnotateArea
            data-testid='annotate-component'
          paths={paths}
          highlightColor={highlightColor.rgbaColor}
            penColor={penColor.rgbaColor}
            onChange={setPaths}
            imageDimensions={imageDimensions}
            backgroundImagePath={screenSnippetPath}
            chosenTool={chosenTool}
            annotateAreaDimensions={annotateAreaDimensions}
          />
        </div>
      </main>
      <footer>
        <button
          data-testid='done-button'
          className='DoneButton'
          onClick={done}>
          {i18n.t('Done', SNIPPING_TOOL_NAMESPACE)()}
        </button>
      </footer>
    </div >
  );
};

export default SnippingTool;
