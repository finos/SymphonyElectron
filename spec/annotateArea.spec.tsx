import { mount } from 'enzyme';
import * as React from 'react';
import AnnotateArea from '../src/renderer/components/annotate-area';
import { Tool } from '../src/renderer/components/snipping-tool';
import * as analyticsHandler from './../src/app/analytics-handler';

const defaultProps = {
    paths: [],
    highlightColor: 'rgba(233, 0, 0, 0.64)',
    penColor: 'rgba(38, 196, 58, 1)',
    onChange: jest.fn(),
    imageDimensions: { width: 800, height: 800 },
    annotateAreaDimensions: { width: 800, height: 800 },
    chosenTool: Tool.pen,
    screenSnippetPath: 'very-nice-path',
};

afterEach(() => {
    jest.clearAllMocks();
});

describe('<AnnotateArea/>', () => {
    it('should render correctly', () => {
        const wrapper = mount(<AnnotateArea {...defaultProps} />);
        expect(wrapper).toMatchSnapshot();
    });

    it('should call onChange when drawn on annotate area', () => {
        const wrapper = mount(<AnnotateArea {...defaultProps} />);
        const area = wrapper.find('[data-testid="annotate-area"]');
        area.simulate('mousedown', { pageX: 2, pageY: 49 });
        area.simulate('mouseup');
        expect(defaultProps.onChange).toHaveBeenCalledTimes(1);
    });

    it('should call onChange with correct pen props if drawn drawn on annotate area with pen', () => {
        const wrapper = mount(<AnnotateArea {...defaultProps} />);
        const area = wrapper.find('[data-testid="annotate-area"]');
        area.simulate('mousedown', { pageX: 2, pageY: 49 });
        area.simulate('mouseup');
        expect(defaultProps.onChange).toHaveBeenCalledWith([{
            color: 'rgba(38, 196, 58, 1)',
            key: 'path0',
            points: [{ x: 0, y: 0 }],
            shouldShow: true,
            strokeWidth: 5,
        }]);
    });

    it('should call onChange with correct highlight props if drawn drawn on annotate area with highlight', () => {
        const highlightProps = {
            paths: [],
            highlightColor: 'rgba(233, 0, 0, 0.64)',
            penColor: 'rgba(38, 196, 58, 1)',
            onChange: jest.fn(),
            imageDimensions: { width: 800, height: 800 },
            annotateAreaDimensions: { width: 800, height: 800 },
            chosenTool: Tool.highlight,
            screenSnippetPath: 'very-nice-path',
        };
        const wrapper = mount(<AnnotateArea {...highlightProps} />);
        const area = wrapper.find('[data-testid="annotate-area"]');
        area.simulate('mousedown', { pageX: 2, pageY: 49 });
        area.simulate('mouseup');
        expect(highlightProps.onChange).toHaveBeenCalledWith([{
            color: 'rgba(233, 0, 0, 0.64)',
            key: 'path0',
            points: [{ x: 0, y: 0 }],
            shouldShow: true,
            strokeWidth: 28,
        }]);
    });

    it('should render path if path is provided in props', () => {
        const pathProps = {
            paths: [{
                points: [{ x: 0, y: 0 }],
                shouldShow: true,
                strokeWidth: 5,
                color: 'rgba(233, 0, 0, 0.64)',
                key: 'path0',
            }],
            highlightColor: 'rgba(233, 0, 0, 0.64)',
            penColor: 'rgba(38, 196, 58, 1)',
            onChange: jest.fn(),
            imageDimensions: { width: 800, height: 800 },
            annotateAreaDimensions: { width: 800, height: 800 },
            chosenTool: Tool.highlight,
            screenSnippetPath: 'very-nice-path',
        };
        const wrapper = mount(<AnnotateArea {...pathProps} />);
        expect(wrapper.find('[data-testid="path0"]').exists()).toEqual(true);
    });

    it('should not render any path if no path is provided in props', () => {
        const wrapper = mount(<AnnotateArea {...defaultProps} />);
        expect(wrapper.find('[data-testid="path0"]').exists()).toEqual(false);
    });

    it('should call onChange with hidden path if clicked on path with tool eraser', () => {
        const pathProps = {
            paths: [{
                points: [{ x: 0, y: 0 }],
                shouldShow: true,
                strokeWidth: 5,
                color: 'rgba(233, 0, 0, 0.64)',
                key: 'path0',
            }],
            highlightColor: 'rgba(233, 0, 0, 0.64)',
            penColor: 'rgba(38, 196, 58, 1)',
            onChange: jest.fn(),
            imageDimensions: { width: 800, height: 800 },
            annotateAreaDimensions: { width: 800, height: 800 },
            chosenTool: Tool.eraser,
            screenSnippetPath: 'very-nice-path',
        };
        const wrapper = mount(<AnnotateArea {...pathProps} />);
        const path = wrapper.find('[data-testid="path0"]');
        path.simulate('click');
        expect(pathProps.onChange).toHaveBeenCalledWith([{
            color: 'rgba(233, 0, 0, 0.64)',
            key: 'path0',
            points: [{ x: 0, y: 0 }],
            shouldShow: false,
            strokeWidth: 5,
        }]);
    });

    it('should send capture_taken BI event on component mount', () => {
        const spy = jest.spyOn(analyticsHandler.analytics, 'track');
        const expectedValue = { action_type: 'annotate_added', element: 'ScreenSnippet' };
        const wrapper = mount(<AnnotateArea {...defaultProps} />);
        const area = wrapper.find('[data-testid="annotate-area"]');
        area.simulate('mousedown', { pageX: 2, pageY: 49 });
        area.simulate('mouseup');
        expect(spy).toBeCalledWith(expectedValue);
    });
});
