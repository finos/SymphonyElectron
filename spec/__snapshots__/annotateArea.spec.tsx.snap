// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AnnotateArea/> should render correctly 1`] = `
<AnnotateArea
  annotateAreaDimensions={
    Object {
      "height": 800,
      "width": 800,
    }
  }
  chosenTool="PEN"
  highlightColor="rgba(233, 0, 0, 0.64)"
  imageDimensions={
    Object {
      "height": 800,
      "width": 800,
    }
  }
  onChange={[MockFunction]}
  paths={Array []}
  penColor="rgba(38, 196, 58, 1)"
  screenSnippetPath="very-nice-path"
>
  <div
    id="annotate-wrapper"
    style={
      Object {
        "height": 800,
        "width": 800,
      }
    }
  >
    <svg
      data-testid="annotate-area"
      height={800}
      id="annotate-area"
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseMove={[Function]}
      onMouseUp={[Function]}
      style={
        Object {
          "cursor": "crosshair",
        }
      }
      width={800}
    />
  </div>
</AnnotateArea>
`;
